client/core/exchangeratefetcher.go

```diff
@@ -0,0 +1,254 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+const (
+	// DefaultFiatCurrency is the currency for displaying assets fiat value.
+	DefaultFiatCurrency = "USD"
+	// fiatRateRequestInterval is the amount of time between calls to the exchange API.
+	fiatRateRequestInterval = 5 * time.Minute
+	// fiatRateDataExpiry : Any data older than fiatRateDataExpiry will be discarded.
+	fiatRateDataExpiry = 60 * time.Minute
+
+	// Tokens. Used to identify fiat rate source.
+	messari       = "Messari"
+	coinpaprika   = "Coinpaprika"
+	dcrdataDotOrg = "dcrdata"
+)
+
+var (
+	dcrDataURL     = "https://explorer.dcrdata.org/api/exchangerate"
+	coinpaprikaURL = "https://api.coinpaprika.com/v1/tickers/%s"
+	messariURL     = "https://data.messari.io/api/v1/assets/%s/metrics/market-data"
+	btcBipID, _    = dex.BipSymbolID("btc")
+	dcrBipID, _    = dex.BipSymbolID("dcr")
+)
+
+// fiatRateFetchers is the list of all supported fiat rate fetchers.
+var fiatRateFetchers = map[string]rateFetcher{
+	coinpaprika:   fetchCoinpaprikaRates,
+	dcrdataDotOrg: fetchDcrdataRates,
+	messari:       fetchMessariRates,
+}
+
+// fiatRateInfo holds the fiat rate and the last update time for an
+// asset.
+type fiatRateInfo struct {
+	rate       float64
+	lastUpdate time.Time
+}
+
+// rateFetcher can fetch fiat rates for assets from an API.
+type rateFetcher func(context context.Context, logger dex.Logger, assets map[uint32]*SupportedAsset) map[uint32]float64
+
+type commonRateSource struct {
+	lastRequest time.Time
+	fetchRates  rateFetcher
+
+	mtx       sync.RWMutex
+	fiatRates map[uint32]*fiatRateInfo
+}
+
+// isExpired checks the last update time for all fiat rates against the
+// provided expiryTime.
+func (source *commonRateSource) isExpired(expiryTime time.Duration) bool {
+	source.mtx.RLock()
+	defer source.mtx.RUnlock()
+	var expiredCount int
+	for _, rateInfo := range source.fiatRates {
+		if time.Since(rateInfo.lastUpdate) > expiryTime {
+			expiredCount++
+		}
+	}
+	totalFiatRate := len(source.fiatRates)
+	return expiredCount == totalFiatRate && totalFiatRate != 0
+}
```

Comment:
Embrace early returns for readability and simplicity:

```diff
 // isExpired checks the last update time for all fiat rates against the
-// provided expiryTime.
+// provided expiryTime. This only returns true if all rates are expired.
 func (source *commonRateSource) isExpired(expiryTime time.Duration) bool {
+       now := time.Now()
        source.mtx.RLock()
        defer source.mtx.RUnlock()
-       var expiredCount int
+       if len(source.fiatRates) == 0 {
+               return false
+       }
        for _, rateInfo := range source.fiatRates {
-               if time.Since(rateInfo.lastUpdate) > expiryTime {
-                       expiredCount++
+               if now.Sub(rateInfo.lastUpdate) < expiryTime {
+                       return false // one not expired is enough
                }
        }
-       totalFiatRate := len(source.fiatRates)
-       return expiredCount == totalFiatRate && totalFiatRate != 0
+       return true
 }
```